### PR TITLE
API Keys - Determinist API Keys + Security changes

### PR DESCRIPTION
**Pull request by gmourier**
🤖  [API Diff](https://github.com/meilisearch/specifications/pull/148#issuecomment-1117158298)
📄 [Rendered](https://github.com/meilisearch/specifications/blob/determinist-api-keys/text/0085-api-keys.md)

---

# Summary

This update to the Keys API makes several changes:

- It is possible to create a deterministic `key` value by specifying a `uid` field at creation.
- It is possible to get a key from the `key` field or the `uid` field.
- Specifying a key to update or delete is made with the `uid` field.
- Adds a `name` field to give a human-readable name to ease API key retrieval in a list at the convenience of the user.
- ~~The master key is only dedicated to API keys management, it can't be used for other endpoints. Making Meilisearch more secure by design and thus preventing users from introducing a security vulnerability.~~
- Updating a key only allows the `name` and `description` fields.
- Introduce new actions to manage API Keys (`keys.get`, `keys.create`, `keys.update`, `keys.delete`). The Default Admin API Key is able to manage keys by default.
- `apiKeyPrefix` tenant token claim is renamed `apiKeyUid`.

---

# Changes

## `uid` Field Addition

The generation of a randomized prefix of 8 characters is removed by the existence of the `uid` field. This field accepts a string representing a UUID v4.

It is possible to specify this `uid` field when creating a key, if not specified it is generated by Meilisearch. 

As the value of `key` is a hash of the `uid` and the master key, it is possible to have deterministic `key` values between instances sharing the same configuration thus keeping some interesting security features as: 

- If the master key changes, all `key` values are automatically changed.
- The API key resources are propagated into the dumps and snapshot without displaying the `key` field in clear.

A key is retrieved from its `uid` rather than its `key` value. e.g. `GET /keys/:uid_or_key` instead of `GET /keys/:key`.

Specifying a `uid` that already exists returns a `409 Conflict`. 

## `name` Field Addition

A `name` field is added to allow a human-readable name to better read the information within a list of keys. The `description` field is kept to add additional information if needed.

For example, the `description` field is useful to give important information about the keys generated by default that would not necessarily have its place in the `name`. This makes it easier to parse in a dashboard when building a UI or in the `GET /keys` response.

The default generated API keys are updated to reflect this change. 

```json
{
    ...,
    "name": "Default Admin API Key",
    "description": "Use it for all other than search operations. Caution! Do not expose it on a public frontend",
    ...
}
```

> Note that the description is updated. `(` / `)` are removed.

## ~~Master key~~

~~**The master key can only be used to manage API keys.**~~

~~Several times, it has been found that users use the master key to make client-side requests, which is problematic for their security. In order to ensure that they do not inadvertently put themselves in danger, the master key can no longer access routes other than `/keys`.~~

## `keys.*` actions

The specification adds new rights dedicated to the management of API keys.

These actions are `keys.get`, `keys.create`, `keys.update`, `keys.delete`.

The default Admin key also has these rights. `actions` is equal to `*` at its creation.

The goal is to give the possibility to users to create dedicated keys for that purpose and to reduce the use of the master key for this endpoint.

## API Key Update

It is no longer possible to update the sensitive information of an API key which could introduce undesired permissions escalation as a side effect.

However, it is possible to update the `name` and `description` fields.

# Out Of Scope

Adding a unique human-readable field as an `alias` to retrieve an API Key other than its `uid` is not considered in this iteration. It should be easy to add it later without breaking changes.

---

# 📡 Attention To Reviewers 📡

@meilisearch/cloud-team We basically need your attention on all these changes. 

The big question mark could be the update of the permissions of an API key on your side, which users could request. This could be done within a transaction on your side to allow the update if it is a vital need. 

The workaround would be to delete the key that needs to be updated and recreate one with the new permissions by giving it the `uid` of the previous one. This way, the `key` value used by clients to authorize requests is unchanged. 

The regeneration of a key would be done in the same way by creating a key copying the permissions but giving it a new `uid`.

Also, do you have any particular constraints concerning the master key, which is now restricted to endpoint `/keys` access only?

@nicolasvienot The addition of the `name` field should allow you to have more fluidity in the construction of the UI and thus avoid having to parse the `description` field to display the default keys properly.

---

## Misc

- [x] Update OpenAPI specification file
- [x] Reorganize the specification

### CODE REVIEW

**Review by brunoocasali — COMMENTED** [review]
Awesome work @gmourier 💪

File: text/0085-api-keys.md [hunk 1]
```diff
@@ -238,313 +227,114 @@ Only the master key allows managing the API keys.
...
 - Regenerate a specific `API Key`.
-- Add a generated id field to paginate the list of API Key.
-- Have a filer/dedicated route to fetch expired key if MeiliSearch do not display them by default on `/keys` in the future.
 - Have an "archive" state where manually deleted API Keys can be restored for a certain amount of time.
 - Add rate-limiting per API Key.
 - A restriction on the maximum offset/limit.
 - Add search parameters restrictions for an API Key.
-- Add rfc2822 format expression for `expiredAt` field. e.g. `Wed, 18 Feb 2022 23:16:09 GMT`
\ No newline at end of file
+- Add rfc2822 format expression for `expiredAt` field. e.g. `Wed, 18 Feb 2022 23:16:09 GMT`
```
[the remark this answers is not in the document]
Comment (gmourier) [reply text/0085-api-keys.md:339]: This is just to be able to support more formats and make it easier to handle. It shouldn't be that specific; just say, "Support more formats."

File: open-api.yaml [hunk 2]
```diff
@@ -545,9 +545,13 @@ components:
       title: key
       type: object
       properties:
+        uid:
+          type: string
+          description: A uuid v4 to identify the API Key. If not specified, it's generated by Meilisearch.
```
[the remark this answers is not in the document]
Comment (gmourier) [reply open-api.yaml:589]: Yes, I've just added the error messages in the dedicated specification 😇

File: text/0085-api-keys.md [hunk 3]
```diff
@@ -1,158 +1,194 @@
...
+- API keys must be provided via the `Authorization` header using the bearer method to authorize a request.
+- The value of the `key` field of an API Key is generated from its `uid` and the master key.
+- When a master key is set at Meilisearch first-launch, it generate two pre-configured default `API Key` resources. A `Default Search API Key` authorizing the search action on all indexes and a `Default Admin API Key` authorizing all actions on all indexes (except managing `/keys` resource).
+- If the master-key changes, all `key` field are re-generated.
+- Default API keys can be modified/deleted from the `/keys` endpoints but are not re-created if Meilisearch has already created them.
+- The master key can only be used to manage API keys. Using the master-key for any other request other than on `/keys` returns an authorization error.
+- API keys can have restrictions on which methods can be accessed via an `actions` list; they also `expiresAt` a specific date time and are restricted to a specific set of `indexes`.
+- `name` and `description` fields are the only editable fields of an API key.
+- API key resources are propagated to snapshots and dumps.
```
[the remark this answers is not in the document]
Comment (gmourier) [reply text/0085-api-keys.md:33]: "`name` and `description` fields are the only editable fields of an API key." was already here!

File: text/0085-api-keys.md [hunk 4]
```diff
@@ -1,158 +1,194 @@
...
 
-```json
+```jsonc
 {
-    "description": "Default Search API Key (Use it to search from the frontend)",
-    "key": "0a6e572506c52ab0bd6195921575d23092b7f0c284ab4ac86d12346c33057f99", //example
+    "uid": "01b4bc42-eb33-4041-b481-254d00cce834", //auto-generated value
+    "key": "0a6e572506c52ab0bd6195921575d23092b7f0c284ab4ac86d12346c33057f99", //auto-generated value
+    "name": "Default Search API Key",
+    "description": "(Use it to search from the frontend)",
```
[the remark this answers is not in the document]
Comment (gmourier) [reply text/0085-api-keys.md:78]: Yes! Thanks @brunoocasali !

**Comment by gmourier** [comment]
@meilisearch/integration-team There remains a design question regarding the JWT payload for tenant token.

Currently, the JWT requires an `apiKeyPrefix`. It should be renamed since the prefix won't exist anymore. Here is 2 possibility:

- `apiKey`: Fill in the signing API key's full `key` field.
- `apiKeyUid`: Fill in the `uid` for the signing API key.

What will be the best choice according to you?

**Comment by gmourier** [comment]
@ManyTheFish will explore two additional if the time permits it. Otherwise it will be considered in the future.

## GET an API Key helper

See if it's possible to have `GET /keys/:uid_or_key` to transparently get the details of an API Key given the field given (`uid` or `key`). If added in the future, it's not a breaking change.

## Add keys management rights

Since the `master-key` will only be used to manage `/keys`. We will explore if we can easily add new rights to manage `keys`. This way, the master-key will be less shared and specific keys could be create to manage the `keys`.

File: text/0061-error-format-and-definitions.md [hunk 5]
```diff
@@ -122,6 +122,80 @@ HTTP Code: `400 Bad Request`
...
+
+This error happens when the `uid` field for an `API Key` resource is invalid.
+
+### Error Definition
+
+HTTP Code: `400 Bad Request`
+
+```json
+{
+    "message": "`uid` field value `:value` is invalid. It should be a valid uuidv4 string or ommited.",
```
Comment (ManyTheFish) [inline text/0061-error-format-and-definitions.md:163]: related to https://github.com/meilisearch/meilisearch/pull/2438#discussion_r885546155

```suggestion
    "message": "`uid` field value `:value` is invalid. It should be a valid UUID v4 string or omitted.",
```

File: text/0061-error-format-and-definitions.md [hunk 6]
```diff
@@ -122,6 +122,80 @@ HTTP Code: `400 Bad Request`
...
+
+This error happens when a user tries to create an API Key that already exists for the given `uid`.
+
+### Error Definition
+
+HTTP Code: `409 Conflict`
+
+```json
+{
+    "message": "`uid` field value `:value` already exists for an API key.",
```
Comment (ManyTheFish) [inline text/0061-error-format-and-definitions.md:139]: related to https://github.com/meilisearch/meilisearch/pull/2438#discussion_r885547329

```suggestion
    "message": "`uid` field value `:value` is already an existing API key.",
```

**Review by ManyTheFish — CHANGES_REQUESTED** [review]

**Comment by gmourier** [comment]
📡

@meilisearch/cloud-team The change about the master key has been discarded. See https://github.com/meilisearch/product/discussions/443#discussioncomment-2862683

@meilisearch/core-team @meilisearch/docs-team @meilisearch/integration-team

- The specification is up-to-date to mention the last successful exploration point made last week.
    - New `actions` are introduced to the management of the API Keys resource. `keys.get`, `keys.create`, `keys.update`, `keys.delete`.
    - `GET` - `/keys/:key` is now `/keys/:uid_or_key`. An API Key can now be fetched by its `uid` or its `key` field value.
    - The tenant token claim `apiKeyPrefix` is renamed `apiKeyUid` and must be given the signing API Key `uid` field value.

**Review by ManyTheFish — CHANGES_REQUESTED** [review]

File: text/0061-error-format-and-definitions.md [hunk 7]
```diff
@@ -122,6 +122,105 @@ HTTP Code: `400 Bad Request`
...
+
+HTTP Code: `400 Bad Request`
+
+```json
+{
+    "message": "`:fieldName` cannot be modified for the given resource.",
+    "code": "immutable_field",
+    "type": "invalid_request",
+    "link":"https://docs.meilisearch.com/errors#immutable_field"
+}
```
[the remark this answers is not in the document]
Comment (gmourier) [reply text/0061-error-format-and-definitions.md:143]: Good idea @ManyTheFish! Shouldn't it in this order? e.g. "`:fieldname` field cannot..."

[the remark this answers is not in the document]
Comment (gmourier) [reply text/0061-error-format-and-definitions.md]: Updated!

File: text/0085-api-keys.md [hunk 8]
```diff
@@ -238,313 +230,117 @@ Only the master key allows managing the API keys.
...
+## 4. Technical Aspects
 
-### 2.1 API Key generation
+### 4.1. API Key generation
 
-A prefix of 8 characters is randomly generated per key and is stored within the instance.
+An `uid` representing by a uuid v4 is generated if not specified at creation by the user.
 
-The final key is then a SHA-2556 hash made of the prefix and the master-key concatenation.
+The final key is then an SHA-256 hash made of the `uid` and the master key concatenation.
```
Comment (ManyTheFish) [inline text/0085-api-keys.md:310]:
```suggestion
The final key is then an HMAC with the master key, pre-hashed by an SHA-256 algorithm, as the secret and the `uid` as the data. HMAC uses an SHA-256 algorithm internally.
```

File: text/0085-api-keys.md [hunk 9]
```diff
@@ -238,313 +230,117 @@ Only the master key allows managing the API keys.
...
 
-`SHA-256(:randomizedPrefix, :master-key)` gives the final `key` field to use.
+`SHA-256(:uid, :master-key)` gives the final `key` field to use when authorizing requests.
```
Comment (ManyTheFish) [inline text/0085-api-keys.md:312]:
```suggestion
`HMAC(secret: SHA-256(:master-key), data: :uid)` gives the final `key` field to use when authorizing requests.
```

File: open-api.yaml [hunk 10]
```diff
@@ -584,9 +584,13 @@ components:
...
+          example: 01b4bc42-eb33-4041-b481-254d00cce834
         key:
           type: string
-          description: The generated key. Generated by Meilisearch.
+          description: The derived key to use in the Authorization header to authorize requests. Generated by Meilisearch with a SHA-256 hash of the uid and the master key.
```
Comment (ManyTheFish) [inline open-api.yaml:593]:
```suggestion
          description: The derived key to use in the Authorization header to authorize requests. Generated by Meilisearch with an HMAC, using an SHA-256 hash, of the uid and the master key.
```

**Review by ManyTheFish — CHANGES_REQUESTED** [review]
@gmourier, I suggested changes about key derivation.

**Review by ManyTheFish — CHANGES_REQUESTED** [review]
Request changes related to https://github.com/meilisearch/meilisearch/pull/2584

**Pull request merged**